frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,372 @@
+                end,
+            })
+        end,
+        checked_func = function()
+            return self:_checked()
+        end,
+        enabled_func = function()
+            return self:_enabled()
+        end,
+        separator = true,
+    }
+end
+
+
+-- add a button for dictquicklookup
+function ReaderUserHyph:addButton(buttons, pos, word, parentWidget)
+    if word:find("[ ,;-%.\n]") then return end
+
+    if G_reader_settings:readSetting("hyph_user_dict") then
+        table.insert(buttons, pos, {
+            {
+                id = "add_user_dict_entry",
+                text = T(_("Modify hyphenation of \"%1\"."), word),
+                callback = function()
+                    if parentWidget then
+                        UIManager:close(parentWidget)
+                    end
+                    self:modifyUserEntry(word)
+                end,
+            },
+        })
+    end
+end
+
+-------------------------------------------
+-- Helper functions for dictionary entries
+-------------------------------------------
+
+-- checks if suggestion is well formated
+function ReaderUserHyph:checkHyphenation(suggestion, word)
+    if suggestion:find("%-%-") then
+        return false -- two or more consecutive '-'
+    end
+
+    suggestion = suggestion:gsub("-","")
+    if Utf8:lower(suggestion) == Utf8:lower(word) then
+        return true -- characters match (case insensitive)
+    end
+    return false
+end
+
+-- this formats a suggestion with upper/lowercase as given in word
+-- Example suggestion=buß-jäger; word=BUẞJäger
+-- return: BUẞ-Jäger
+-- notice the 'ß' is coded in two bytes, whereas 'ẞ' uses three bytes. (terrible mapping)
+function ReaderUserHyph:formatHyphenation(suggestion, word)
+    if not suggestion then return word end
+
+    local suggestion_lower = Utf8:lower(suggestion)
+    local word_lower = Utf8:lower(word)
+    local val = ""
+    local i, j = 1, 1
+    local word_len = word_lower:len()
+    while i <= word_len do
+        if suggestion:sub(j,j) == "-" then
+            val = val .. "-"
+            j = j + 1
+        end
+        if word_lower:sub(i, i) == suggestion_lower:sub(j, j) then
+            -- e.g.: lowercase ß (Utf8: 0xC3 0x9F), uppercase ẞ (Utf8: 0xE1 0xBA 0x9E)
+            -- so a matching lowercase must be translated to uppercase if necessary
+            if word_lower:sub(i, i) == word:sub(i,i) then -- word[i] is lowercase, then dumb copy
+                val = val .. word:sub(i, i) -- use uppercases from word, as suggestion is always lowercase
+                i = i + 1
+                j = j + 1
+            else  -- case differs in word[i] and word_lower[i]
+                local multibyte = false
+                for low_c, up_c in pairs(Utf8.lc_uc) do
+                    if (word_lower:sub(i):find(low_c) or 0) == 1 then -- Utf8-multibyte at position i
+                        print("multibyte ".. low_c .. " " .. up_c)
```

Comment:
I hope you'll be able to avoid all this with utf8proc_lower & co.
If not, might be easier to work with unicode codepoints around here ? We have a few functions in frontend/util.lua or base/ffi/util.lua that might help (or not)

Reply:
I am on the way for that!!!!